// flatiron-cli-users/lib/properties.js
/*
 * properties.js: Properties for the prompts in flatiron-cli-users.
 *
 * (C) 2010, Nodejitsu Inc.
 *
 */
 
module.exports = {
  email: {
    name: 'email',
    validator: /^((([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+(\.([a-z]|\d|[!#\$%&'\*\+\-\/=\?\^_`{\|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+)*)|((\x22)((((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(([\x01-\x08\x0b\x0c\x0e-\x1f\x7f]|\x21|[\x23-\x5b]|[\x5d-\x7e]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(\\([\x01-\x09\x0b\x0c\x0d-\x7f]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))))*(((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(\x22)))@((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.?$/i,
    warning: 'Must be a valid email address'
  },
  password: {
    name: 'password',
    hidden: true
  },
  'set password': {
    name: 'set password',
    hidden: true
  },
  'confirm password': {
    name: 'confirm password',
    hidden: true
  },
  username: {
    name: 'username',
    validator: /^[\w|\-|\.]+$/,
    warning: 'Username can only be letters, numbers, dots and dashes'
  },
  reset: {
    name: 'request password reset',
    validator: /y[es]?|n[o]?/,
    warning: 'Must respond yes or no',
    default: 'no'
  },
  login: {
    name: 'login',
    message: '(yes/no)',
    validator: /^y[yes]+|n[no]+/,
    warning: 'Must answer yes or no.',
    default: 'yes'
  }
};
